### DISCUSSION TITLE
GDML Arrays into Geant4

### DISCUSSION THREAD

**Discussion by MITgnome** [KeithSloan/GDML#106] [General]
Would you expect the Geant4 GDML Reader to properly import an gdml _**array**_ of parts. The gdml files that I have succuessfully imported into Geant4 all seem to not be hierarchical. Making Arrays of detectors would make life easier for building the geometry, but I want to make sure that Geant4 would be able to read these types of gdml files in.
<img width="1642" alt="Screenshot 2023-03-12 at 11 32 27 AM" src="https://user-images.githubusercontent.com/1777892/224555303-01918d57-6d1f-41bd-81c9-c881b01583ba.png">

**Comment by mhindi2** [comment]
I am not sure I understand the question. The GDML language does not istelf have an array tag. What it does have is a loop tag that allows the placement of arrays of object using loop parametrs. At present the GDML workbench does not make use of the loop tag to export arrays. However, arrays such as the one shawn in your attached figure are exported properly as multiple individual parts placed correctly.  All elements of the arrays share the same material.  That structure should be read correctly into geant4. Check the examples provided with the GDML Workbench Wiki. Perhaps I am missing something about the question.

**Comment by KeithSloan** [comment]
Small point. In the screen shot the Part(Volume) experiment is at the same level as the worldVOL. The exporter should handle this but if you were to hand code the GDML there would only be one world Volume at the root level. If you are seeing a problem, my I suggest you select the Part/Volume experiment and drag it to the worldVol and this would then make the structure the same as if hand coded. Again if you are seeing a problem with this please let us know so we can address.

**Comment by mhindi2** [comment]
Yes, the experiment volume should be dragged onto the World volume, and then the World volume box be made large enough to encompass the experimet. There is an icon in the GDML Workbench that adjusts the size of the box to encompass all the parts underneath the World volume.

**Comment by mhindi2** [comment]
@MITgnome, KeithSloan finally explained the problem to me, which, I am afraid I missed on first viewing. Currently the GDML workbench _does not_ support the export an array of logical voulmes (i.e., your DetectorModule and anything underneath it). Currently the GDML workbench only exports an array of _solids_, not logical volumes. The array is exported as a multiunion, which is a _single_ solid, with a _single_ material. Exporting the structure in your figure would involve an update, which may happen in the future.

_However_, there _is_ a way to accomplish the structure you have, albeit somewhat less elegantly:
Instead of having an array of assemblies, you can have an assembly of arrays. The attached screenshot shows how you would effectively accomplish the same geometry:
![image](https://user-images.githubusercontent.com/95139464/224767727-ceaa17df-6684-4605-95f3-9cf8b4e19762.png)


This is slightly more work, in that _four_ arrays (each consisting of one _solid_) have to be created instead of _one_ array containing _four_ logical volumes, and one has to make sure each of the arrays contains the same number of elements and displacements. Please note that the arrays _have_ to be arrays of solids, not logical volumes.

**Comment by KeithSloan** [comment]
Have created an update for an Orthogonal Array of a Part.

Need to create some more more complicated test cases i.e. Arrray of Array and more Complicated Part to do some more testing

[testExperiment.FCStd.txt](https://github.com/KeithSloan/GDML/files/10984102/testExperiment.FCStd.txt)
[testExperiment-worldVOL.gdml.txt](https://github.com/KeithSloan/GDML/files/10984103/testExperiment-worldVOL.gdml.txt)
![Image 15-03-2023 at 19 52](https://user-images.githubusercontent.com/2291247/225429463-d32cb554-0890-4613-82cf-c748bf2f6b53.jpg)

Once this is all okay, I will move onto other types of Array with a Part/Volume

**Comment by mhindi2** [comment]
Hi Keith,

First off, I am not sure whom this reply is going to.

I pulled Main into one of my local repositories and started testing. I 
began with the polarArray test and reproduced what you had.

(1) The test you had should not be given as an example (in the github), 
because it produces overlapping volumes which geant complains about.

(2) When I tried to move (a) either the LV containing the box or (b) the 
GDML_box itself, (so as to produce non-overlapping volumes), that 
placement was not reflected in the output gdml file, see attached 
screenshot and used FreeCAD file.

So, the placements of the array elements should be convolved 
(multiplied) by the placements of the underlying elements, as I do for 
other assembly objects. That has to be done recursively (as I do it), so 
that no matter the level of daughters, the overall placement is still 
correct.

I have not tested the rectangular array implementation yet, but I 
suspect it will suffer from the same shortcoming, i.e., a rotation of 
the underlying element, will probably be not reflected in the array.

Munther

**Comment by mhindi2** [comment]
@MITgnome 
There is an update to the Main branch in the GitHub that should allow the export of the original structure you first posted. Please give it a try an let us know if it works or not.
